Machine Learning - Autumn assessment

Tasks

1. Download the abalone.csv file and save to your repository. 

I downloaded the file from the link provided & with my CMD commited the changes to me repo to Github Repo. 

2. Use the read_csv() function from the pandas library to load the data into your notebook.

Firstly we need to add in all of our libaries in order to complete all the requests in tasks 1 through 5. Brief discription of what each libary does below.

Pandas: Is a python package for data manipulation and analysis in Python, it's very good for working with with "relational" or "labeled" data, this makes it useful as the Abalone dataset is a labeled dataset. Also most importantly it allows us to actually load in the Abalone CSV.
https://pypi.org/project/pandas/

NumPy: Is a libary which we will mainly use for for point 4 of our tasks which will involve some high level mathematical functions. Numpy is also important as it will help us dealing with multi dimensional arrays within the Abalone dataset
https://www.codecademy.com/article/introduction-to-numpy-and-pandas
https://www.kaggle.com/code/chats351/introduction-to-numpy-pandas-and-matplotlib


Seaborn: Is a phython library used for data visualisation which we will be using in point 5 of our tasks. It is based on the the Matplotlib libarary and makes the visualization  aspect of our plots and graphs. 
https://seaborn.pydata.org/
https://www.geeksforgeeks.org/introduction-to-seaborn-python/

Matplotlib: Is a library used for creating interative visualisations for our Abalone dataset, this libarary will be used mostly in point 5 of our tasks & integrates very nicely with Seaborn to create interactive and highqulaity visualisations.
https://matplotlib.org/

In [80]:
# importing pandas & Numpy file
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [81]:
# Load the data
df = pd.read_csv('abalone.csv')

# Display the first few rows so I know I have loaded correctly
df.head()

sex  length  diameter  height  whole_weight  shucked_weight  viscera_weight  \
0   M   0.455     0.365   0.095        0.5140          0.2245          0.1010   
1   M   0.350     0.265   0.090        0.2255          0.0995          0.0485   
2   F   0.530     0.420   0.135        0.6770          0.2565          0.1415   
3   M   0.440     0.365   0.125        0.5160          0.2155          0.1140   
4   I   0.330     0.255   0.080        0.2050          0.0895          0.0395   

   shell_weight  rings  
0         0.150     15  
1         0.070      7  
2         0.210      9  
3         0.155     10  
4         0.055      7

3. Within the notebook, discuss the classification of each variable in the dataset based on common variable types and measurement scales in mathematics, statistics, and Python

Variable types are broken down into two categories. Categorical & Numeric. 
Categorical data is then broken into two sub types which are Nominal and Ordinal while for numerical we have Discrete and continous.
First we will look at the Categorical variables, what they are & if present in the Abalone dataset. https://www150.statcan.gc.ca/n1/edu/power-pouvoir/ch8/5214817-eng.htm

Categorical variables

Are variables which have two or more categories that represents distinct categories or groups.
A Binary question such as Yes/No is a categorical variable as a basic example. With Categorical variables there is no intrinsic ordering to the categories.
Visually categorical data is best represented in bar charts & histograms for example. https://stats.oarc.ucla.edu/other/mult-pkg/whatstat/what-is-the-difference-between-categorical-ordinal-and-interval-variables/

The two sub types of categorical data are Nominal Variables & Ordinal Variables

Nominal Variables are when categories have no order or rank. For example gender (Male/Female) or hair color (Black/Brown/Blonde).
In our Abalone dataset there is only one Nominal variable which is the sex column (Male/female/Infant) https://www.cuemath.com/data/nominal-variable/

Ordinal Variables are categorical variables where by the categories do have a rank or order designated to them. For example frequency of exercise being Often,Sometimes,rarely & never. These values provide an order but we can't calculate the exact distance between rarely & sometimes. https://www.scribbr.com/statistics/ordinal-data/
In our Abalone dataset there is no variable which falls under this category.

Numerical Variables

Are a variable which has a value where you can describe a measurable quantity or numerical meaning. https://socratic.org/questions/what-is-a-numerical-variable-and-what-is-a-categorical-variable
For example how many ML's in a cup of water would be a numerical variable. 

The two sub types of Numerical Variables are Continuous Variables & Discrete Variables:

Continuous Variables are variables which can take any value within a range example being the cup of water again.
A continuous variable takes on an infinite number of possible values within a given range. https://articles.outlier.org/discrete-vs-continuous-variables#section-what-is-a-continuous-variable

In our Abalone dataset the following variables are Continous: Length, Diameter, Height, Whole Weight, Shucked Weight, Viscera Weight & Shell Weight.

Discrete Variable a variable that takes on distinct, countable values. For example the number of steps you take in a day is a discrete variable.
https://articles.outlier.org/discrete-vs-continuous-variables#section-what-is-a-discrete-variable

In Abalone the only discrete variable is the Ring variable, Abalones age is counted by the number of rings present hence it being a discrete variable.


4. Select, demonstrate, and explain the most appropriate summary statistics to describe each variable.

In [82]:
# Summary statistics for categorical variable 'Sex'
sex_counts = df['sex'].value_counts() # count of the unique value for the 3 sex's
sex_proportions = df['sex'].value_counts(normalize=True) # relative frequencey of each unique value for the 3 sex's

print("Frequency Counts for 'sex':") # Just printing text
print(sex_counts) # displays the count of each Sex
print("\nProportions for 'sex':") # Just printing text
print(sex_proportions) # displays the count of each Sex

Frequency Counts for 'sex':
M    1528
I    1342
F    1307
Name: sex, dtype: int64

Proportions for 'sex':
M    0.365813
I    0.321283
F    0.312904
Name: sex, dtype: float64


With the above code we just show the basic count of each sex (Male, Female, Infant) & the proportions code gives us a view into the percentage distribution throughout the 3 different types of sexes. From this we can see the majority is Male & that male has the highest percentage distrubtion while Female is the lowest with the lowest percentage distrubtion in the dataset.

In [83]:
# Summary statistics for numerical variables
numerical_stats = df[['length', 'diameter', 'height', 'whole_weight', 'shucked_weight', 'viscera_weight', 'shell_weight', 'rings']].describe() # defining the numbercial stats & using the describe function


print("Summary Statistics for Numerical Variables:") # Printing text
print(numerical_stats) #printing the output from numerical_stats


Summary Statistics for Numerical Variables:
            length     diameter       height  whole_weight  shucked_weight  \
count  4177.000000  4177.000000  4177.000000   4177.000000     4177.000000   
mean      0.523992     0.407881     0.139516      0.828742        0.359367   
std       0.120093     0.099240     0.041827      0.490389        0.221963   
min       0.075000     0.055000     0.000000      0.002000        0.001000   
25%       0.450000     0.350000     0.115000      0.441500        0.186000   
50%       0.545000     0.425000     0.140000      0.799500        0.336000   
75%       0.615000     0.480000     0.165000      1.153000        0.502000   
max       0.815000     0.650000     1.130000      2.825500        1.488000   

       viscera_weight  shell_weight        rings  
count     4177.000000   4177.000000  4177.000000  
mean         0.180594      0.238831     9.933684  
std          0.109614      0.139203     3.224169  
min          0.000500      0.001500     1.000000 

We use describe() to give us a list/summary of statistics related to all of our numerical data. Lets look at each row.

Count: this just gives us a count of the number of values within the dataset, same as when we looked at the counts of each sex the value is 4177 which is equal to all the sexes of Abalone sexes added together.

Mean: Gives us an average value of length, diameter, height etc for all the Abalone in the dataset

Std: This is standard deviation which is a measure of how dispersed the data is in relation to the mean. Low, or small, standard deviation indicates data are clustered tightly around the mean, and high, or large, standard deviation indicates data are more spread out. We can see from this there is a low std across the board for all numerical variables except for Rings.

Min: This is just the smallest value for each column in the dataset, pretty self explantory.

25%: 25th, first quartile, value below which 25% of the data falls.

50%: 50th percentile, middle value of the data.

75%: 75th, third quartile, value below which 75% of the data falls.

max: This is just the smaximum value for each column in the dataset, pretty self explantory.

5. Select, demonstrate, and explain the most suitable plot(s) for each variable.